File: test/asaga_sparse.cpp
#include "ipp.ipp"
#include "statick/solver/asaga.hpp"
constexpr size_t N_ITER = 200;
size_t THREADS = mkn::kul::cpu::threads() < 12 ? mkn::kul::cpu::threads() : 12;
int main() {
  using T = double;
  using FEATURES = statick::Sparse2D<T>;
  using LABELS = statick::Array<T>;
  using MODEL = statick::ModelLogReg<std::shared_ptr<FEATURES>, std::shared_ptr<LABELS>>;
  using PROX = statick::ProxL2Sq<T>;
  using SOLVER = statick::ASAGA;
  using SODAO = SOLVER::DAO<MODEL, statick::solver::History<T, statick::solver::Tolerance<T>>>;
  auto SOLVE = &SOLVER::template SOLVE<SODAO, PROX>;
  MODEL::DAO modao(FEATURES::FROM_CEREAL("url.features.cereal"),
                   LABELS::FROM_CEREAL("url.labels.cereal"));
  const T STRENGTH = (1. / modao.n_samples()) + 1e-10;
  SODAO dao(modao, N_ITER, modao.n_samples(), THREADS);
  PROX prox(STRENGTH);
  dao.step = 0.00257480411965l;
  dao.history.tol = 1e-5;
  SOLVE(dao, modao, prox);
  std::vector<T> &objs(dao.history.objectives);
  auto min_objective = *std::min_element(std::begin(objs), std::end(objs));
  KLOG(INF) << min_objective;
  auto history = dao.history.time_history;
  auto log_every_n_epochs = dao.history.log_every_n_epochs;
  for (size_t i = 1; i < objs.size(); i++) {
    auto log_dist = objs[i] == min_objective ? 0 : log10(objs[i] - min_objective);
    std::cout << THREADS << " " << i * log_every_n_epochs << " " << history[i] << " "
              << "1e" << log_dist << std::endl;
  }
  return 0;
}

File: test/asaga_sparse.py
import math
from statick.linear_model import ModelLogReg
from statick.prox import ProxL2Sq
from statick.solver import SAGA
import statick

N_THREADS=8
MAX_ITER=200
STEP=0.00257480411965
TOL=1e-5

X = statick.load_double_sparse2d("url.features.cereal")
y = statick.load_double_array("url.labels.cereal")

solver = SAGA(step=STEP, max_iter=MAX_ITER, verbose=False, tol=TOL, n_threads=N_THREADS) \
          .set_model(ModelLogReg().fit(X, y))                                        \
          .set_prox (ProxL2Sq(strength=((1. / X.shape[0]) + 1e-10)))
solver.solve()
objs = solver.objectives
history = solver.time_history
min_objective = min(objs)
for i in range(len(objs)):
    log_dist = 0 if objs[i] == min_objective else math.log10(objs[i] - min_objective);
    print(str(N_THREADS) + " " + str(i * solver.log_every_n_epochs) + " " + \
    	  str(history[i]) + " " + "1e" + str(log_dist));
